llvm/lib/Target/X86/X86FrameLowering.cpp

```diff
@@ -2194,13 +2194,35 @@ void X86FrameLowering::emitPrologue(MachineFunction &MF,
   // flag (DF in EFLAGS register). Clear this flag by creating "cld" instruction
   // in each prologue of interrupt handler function.
   //
-  // FIXME: Create "cld" instruction only in these cases:
+  // Create "cld" instruction only in these cases:
   // 1. The interrupt handling function uses any of the "rep" instructions.
   // 2. Interrupt handling function calls another function.
   //
-  if (Fn.getCallingConv() == CallingConv::X86_INTR)
-    BuildMI(MBB, MBBI, DL, TII.get(X86::CLD))
-        .setMIFlag(MachineInstr::FrameSetup);
+  if (Fn.getCallingConv() == CallingConv::X86_INTR) {
+    bool NeedsCLD = false;
+
+    // Check if the function calls another function.
+    for (const BasicBlock &BB : Fn) {
+      for (const Instruction &I : BB) {
+        if (isa<CallInst>(I)) {
+          NeedsCLD = true;
+          break;
+        }
+
+        // Check for rep opcode.
+        unsigned Opcode = I.getOpcode();
+        if (X86::REP_MOVSB_32 <= Opcode && Opcode <= X86::REP_STOSW_64) {
+          NeedsCLD = true;
+          break;
+        }
```

Comment:
Do we need to worry about inline assembly that contains rep instructions?

Reply:
Should we do it if there is inline assembly at all?

Reply:
Or maybe we can check for rep specifically?

Reply:
> Or maybe we can check for rep specifically?

what if there is inline assembler that has rep AND cld? Then we'd have cld execute twice. Which I guess is not harmful but honestly, I think that people who write inline assembly ought to know what they are doing, and as such perhaps another pass should cleanup duplicate unneeded instructions.

Reply:
Also inlineasm is basically a special call instruction, so in practice this will just make cld anyway

Reply:
inlineasm is a call in IR. It is not a call in MachineIR. It is its own instruction, TargetOpcode::INLINEASM and TargetOpcode::INLINEASM_BR. There is MachineInstr::isInlineAsm to check for these opcodes.

Reply:
Okay, but the main question is that if someone who writes asm knows what they are doing with using rep manually, should we assume they need to have the cld emitted for them? We cannot assume the intention for them like that. For all we know, they could not be expecting cld to happen at all, or they write it themselves.

Reply:
I think that for now, we should just emit cld if we find inline asm. We can figure out how to parse it later. Because there are also things like possible function calls within the inline asm. This would preserve the behavior prior to this patch anyway in these cases.

Reply:
@topperc Thoughts?